From fd87171699bcb96746c48f38d3ee970c1e1ca928 Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Sun, 22 Nov 2015 16:51:38 +0200
Subject: [PATCH 06/38] ARM: i.MX6UL: cl-som-imx6ul: add support for WiLink8

Add support for wilink8 wlan and bluetooth on cl-som-imx6ul.
* enable uart2 used for bluetooth with wilink8 with flow control
* configure usdhc1 to be used for sdio communication to WiLink8 wlan
* configure gpios used for wlan/bt enables and wlan_irq
* setup voltage regulator used for wl8 wlan power control

WiLink8 shares the same mmc bus with the evaluation board mmc slot.
In order to allow mmcboot the WiLink8 enabled device tree file has been created.

Signed-off-by: Valentin Raevsky <valentin@compulab.co.il>
---
 arch/arm/boot/dts/cl-som-imx6ul-wilink.dts | 98 ++++++++++++++++++++++++++++++
 arch/arm/boot/dts/cl-som-imx6ul.dts        | 17 +++++-
 2 files changed, 112 insertions(+), 3 deletions(-)
 create mode 100644 arch/arm/boot/dts/cl-som-imx6ul-wilink.dts

diff --git a/arch/arm/boot/dts/cl-som-imx6ul-wilink.dts b/arch/arm/boot/dts/cl-som-imx6ul-wilink.dts
new file mode 100644
index 0000000..48d44d0
--- /dev/null
+++ b/arch/arm/boot/dts/cl-som-imx6ul-wilink.dts
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2015 CompuLab LTD.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include "cl-som-imx6ul.dts"
+
+/ {
+	regulators {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		wlan_en_reg: wlan_en-reg {
+			compatible = "regulator-fixed";
+			regulator-name = "wlan-en-regulator";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			gpio = <&gpio5 1 GPIO_ACTIVE_HIGH>;
+			/* WLAN card specific delay */
+			startup-delay-us = <70000>;
+			enable-active-high;
+		};
+	};
+
+	kim {
+		compatible = "kim";
+		nshutdown_gpio = <128>;  /* GPIO5_IO00 */
+		dev_name = "/dev/ttymxc1";
+		flow_cntrl = <1>;
+		baud_rate = <3000000>;
+	};
+
+	btwilink {
+		compatible = "btwilink";
+	};
+};
+
+&iomuxc {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_hog_1 &pinctrl_hog_2>;
+	cl-som-imx6ul {
+		pinctrl_hog_2: hoggrp-2 {
+			fsl,pins = <
+				/* Wlan En */
+				MX6UL_PAD_SNVS_TAMPER1__GPIO5_IO01 0x17059
+				/* Wlan IRQ */
+				MX6UL_PAD_SNVS_TAMPER9__GPIO5_IO09 0x17059
+				/* Bt En */
+				MX6UL_PAD_SNVS_TAMPER0__GPIO5_IO00 0x17059
+			>;
+		};
+
+		pinctrl_uart2: uart2grp {
+			fsl,pins = <
+				MX6UL_PAD_UART2_TX_DATA__UART2_DCE_TX 0x1b0b1
+				MX6UL_PAD_UART2_RX_DATA__UART2_DCE_RX 0x1b0b1
+				MX6UL_PAD_UART2_CTS_B__UART2_DCE_CTS 0x1b0b1
+				MX6UL_PAD_UART2_RTS_B__UART2_DCE_RTS 0x1b0b1
+			>;
+		};
+	};
+};
+
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart2>;
+	status = "okay";
+
+	/* enable rts/cts usage on uart2 */
+	fsl,uart-has-rtscts;
+};
+
+&usdhc1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usdhc1>;
+	/* clean up the removable device fields */
+	cd-gpios = <>;
+	wp-gpios = <>;
+	no-1-8-v;
+	keep-power-in-suspend;
+	enable-sdio-wakeup;
+	vmmc-supply = <&wlan_en_reg>;
+	non-removable;
+	cap-power-off-card;
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+	wlcore: wlcore@0 {
+		compatible = "ti,wl1835";
+		reg = <2>;
+		interrupt-parent = <&gpio5>;
+		interrupts = <9 IRQ_TYPE_LEVEL_HIGH>;
+	};
+};
diff --git a/arch/arm/boot/dts/cl-som-imx6ul.dts b/arch/arm/boot/dts/cl-som-imx6ul.dts
index d15b29d..34a66d6 100644
--- a/arch/arm/boot/dts/cl-som-imx6ul.dts
+++ b/arch/arm/boot/dts/cl-som-imx6ul.dts
@@ -9,6 +9,7 @@
 /dts-v1/;
 
 #include <dt-bindings/input/input.h>
+#include <dt-bindings/interrupt-controller/irq.h>
 #include "imx6ul.dtsi"
 
 / {
@@ -92,6 +93,7 @@
 			regulator-max-microvolt = <3300000>;
 			regulator-always-on;
 		};
+
 	};
 
 	leds {
@@ -108,7 +110,7 @@
 &iomuxc {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_hog_1>;
-	imx6ul-som-compulab {
+	cl-som-imx6ul {
 		pinctrl_hog_1: hoggrp-1 {
 			fsl,pins = <
 				/* Heartbeat */
@@ -136,6 +138,15 @@
 			>;
 		};
 
+		pinctrl_uart2: uart2grp {
+			fsl,pins = <
+				MX6UL_PAD_UART2_TX_DATA__UART2_DCE_TX 0x1b0b1
+				MX6UL_PAD_UART2_RX_DATA__UART2_DCE_RX 0x1b0b1
+				MX6UL_PAD_UART2_CTS_B__UART2_DCE_CTS 0x1b0b1
+				MX6UL_PAD_UART2_RTS_B__UART2_DCE_RTS 0x1b0b1
+			>;
+		};
+
 		pinctrl_uart3: uart3grp {
 			fsl,pins = <
 				MX6UL_PAD_UART3_TX_DATA__UART3_DCE_TX 0x1b0b1
@@ -288,8 +299,8 @@
 &usdhc1 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_usdhc1>;
-        cd-gpios = <&gpio1 19 0>;
-        wp-gpios = <&gpio1 18 0>;
+	cd-gpios = <&gpio1 19 0>;
+	wp-gpios = <&gpio1 18 0>;
 	keep-power-in-suspend;
 	enable-sdio-wakeup;
 	status = "okay";
-- 
1.9.1

